flamedisx/xenon/itp_map.py

```diff
@@ -109,58 +122,149 @@ def __init__(self, data):
+            axes = np.roll(np.arange(len(grid) + 1), -1)
+            csys = np.transpose(csys, axes)
+            csys = np.array(csys).reshape((-1, len(grid)))
             self.dimensions = len(grid)
         else:
-            self.dimensions = len(cs[0])
+            csys = np.array(csys)
+            self.dimensions = len(csys[0])
 
-        self.coordinate_system = cs
+        self.coordinate_system = csys
         self.interpolators = {}
         self.map_names = sorted([k for k in self.data.keys()
-                                 if k not in self.data_field_names])
+                                 if k not in self.metadata_field_names])
 
         log = logging.getLogger('InterpolatingMap')
-        log.debug('Map name: %s' % self.data['name'])
+        log.debug('Map name: %s' % self.data.get('name',
+                                                 "NO NAME?!"))
         log.debug('Map description:\n    ' +
-                  re.sub(r'\n', r'\n    ', self.data['description']))
+                  re.sub(r'\n', r'\n    ', self.data.get('description',
+                                                         "NO DESCRIPTION?!")))
         log.debug("Map names found: %s" % self.map_names)
 
         for map_name in self.map_names:
-            map_data = np.array(self.data[map_name])
-            array_valued = len(map_data.shape) == self.dimensions + 1
+            # Specify dtype float to set Nones to nan
+            map_data = np.array(self.data[map_name], dtype=np.float64)
+            if len(self.coordinate_system) == len(map_data):
+                array_valued = len(map_data.shape) == 2
+            else:
+                array_valued = len(map_data.shape) == self.dimensions + 1
+
             if self.dimensions == 0:
                 # 0 D -- placeholder maps which take no arguments
                 # and always return a single value
-                def itp_fun(positions, _data=map_data):
-                    return np.array([_data])
+                def itp_fun(positions):
+                    return np.array([map_data])
+
+            elif method == 'RectBivariateSpline':
+                itp_fun = self._rect_bivariate_spline(csys, map_data, array_valued, **kwargs)
+
+            elif method == 'RegularGridInterpolator':
+                itp_fun = self._regular_grid_interpolator(csys, map_data, array_valued, **kwargs)
+
+            elif method == 'WeightedNearestNeighbors':
+                itp_fun = self._weighted_nearest_neighbors(csys, map_data, array_valued, **kwargs)
+
             else:
-                if array_valued:
-                    map_data = map_data.reshape((-1, map_data.shape[-1]))
-                itp_fun = InterpolateAndExtrapolate(points=np.array(cs),
-                                                    values=np.array(map_data),
-                                                    array_valued=array_valued)
+                raise ValueError(f'Interpolation method {method} is not supported')
```

Comment:
In case of error, could you add the supported methods in the error message?

Reply:
Interpolation is a delicate operation and I think it would be better for the user to safer for the user to be mindful of the exact interpolation algorithm. The error message in `ValueError` could be edited to include the supported interpolation algorithms but that would have to be changed each time an algorithm is included.

The interpolation functions are adapted from those in straxen and we expect to see more changes in the near future.